src/control/control.cpp

```diff
@@ -0,0 +1,351 @@
+/**
+ * @file control.hpp
+ * @author Nilusink
+ * @version 1.0
+ * @date 2023-03-04
+ * 
+ * @copyright Copyright frenchbakery (c) 2023
+ * 
+ */
+#include <kipr/create/create.h>
+#include <nlohmann/json.hpp>
+#include "term_colors.h"
+#include "control.hpp"
+#include <iostream>
+#include <vector>
+#include <memory>
+
+
+using json = nlohmann::json;
+
+// initialize vectors
+std::vector<std::shared_ptr<Control::Motor>> Control::motors; 
+std::vector<std::shared_ptr<Control::Servo>> Control::servos;
+std::vector<std::shared_ptr<Control::Digital>> Control::digitals;
+std::vector<std::shared_ptr<Control::Analog>> Control::analogs;
+
+
+void Control::setup()
+{
+    // initialize motors / servos / digitals / analogs
+    for (int i = 0; i <= N_MOTORS; i++)
+    {
+        motors.emplace_back(std::make_shared<Motor>(i));
+    }
+
+    for (int i = 0; i <= N_SERVOS; i++)
+    {
+        servos.emplace_back(std::make_shared<Servo>(i));
+    }
+
+    for (int i = 0; i <= N_DIGITALS; i++)
+    {
+        digitals.emplace_back(std::make_shared<Digital>(i));
+    }
+
+    for (int i = 0; i <= N_ANALOGS; i++)
+    {
+        analogs.emplace_back(std::make_shared<Analog>(i));
+    }
+}
+
+
+/**
+ * @brief handle all get requests
+ * 
+ * @param data input json object
+ * @param out output json object
+ */
+void handle_get(json data, json &out)
+{
+    std::string request = data["request"];
+
+    // motors
+    if (request == "motor")
+    {
+        if (!data.contains("port"))
+        {
+            out["success"] = false;
+            out["reason"] = "message doesn't contain \"port\" key";
+            return;
+        }
+
+        int port = data["port"];
```

Comment:
You may want to consider using .at("key") instead of ["key"] when READING data from the json object. This will through an exception if the key doesn't exist (which you have checked for already so no big deal usually) where as  ["key"] will simply create the key and give a weird error if the type is unexpected or even worse no error at all and cause unexpected behavior.

Reply:
done in 94deca69203cb88cafb86dba7a3ff2a6f72451a6